cmd/erasure.go

```diff
@@ -0,0 +1,167 @@
+	"io"
+
+	"github.com/klauspost/reedsolomon"
+	"github.com/minio/minio/pkg/bitrot"
+)
+
+// OfflineDisk represents an unavailable disk.
+var OfflineDisk StorageAPI // zero value is nil
+
+// MissingShard indicates a missing data/parity shard for erasure coding
+var MissingShard []byte // zero value is nil
+
+// ErasureFileInfo contains information about an erasure file operation (create, read, heal).
+type ErasureFileInfo struct {
+	Size      int64
+	Algorithm bitrot.Algorithm
+	Keys      [][]byte
+	Checksums [][]byte
+}
+
+// XLStorage represents an array of disks.
+// The disks contain erasure coded and bitrot-protected data.
+type XLStorage struct {
+	disks                    []StorageAPI
+	erasure                  reedsolomon.Encoder
+	dataBlocks, parityBlocks int
+}
+
+// NewXLStorage creates a new XLStorage. The storage erasure codes and protects all data written to
+// the disks.
+func NewXLStorage(disks []StorageAPI, dataBlocks, parityBlocks int) (s XLStorage, err error) {
+	erasure, err := reedsolomon.New(dataBlocks, parityBlocks)
+	if err != nil {
+		return s, Errorf("failed to create erasure coding: %v", err)
+	}
+	s = XLStorage{
+		disks:        make([]StorageAPI, len(disks)),
+		erasure:      erasure,
+		dataBlocks:   dataBlocks,
+		parityBlocks: parityBlocks,
+	}
+	copy(s.disks, disks)
+	return
+}
+
+// AppendWriters creates one writer for every disk of the storage array.
+// Every writer appends data to the corresponding storage disk.
+func (s *XLStorage) AppendWriters(volume, path string) []io.Writer {
+	writers := make([]io.Writer, len(s.disks))
+	for i := range writers {
+		writers[i] = StorageWriter(s.disks[i], volume, path)
+	}
+	return writers
+}
+
+// ErasureEncode encodes the given data and returns the erasure-coded data.
+// It returns an error if the erasure coding failed.
+func (s *XLStorage) ErasureEncode(data []byte) ([][]byte, error) {
+	encoded, err := s.erasure.Split(data)
+	if err != nil {
+		return nil, Errorf("failed to split data: %v", err)
+	}
+	if err = s.erasure.Encode(encoded); err != nil {
+		return nil, Errorf("failed to encode data: %v", err)
+	}
+	return encoded, nil
+}
+
+// ErasureDecodeDataBlocks decodes the given erasure-coded data.
+// It only decodes the data and parity blocks but does not verify them.
+// It returns an error if the decoding failed.
+func (s *XLStorage) ErasureDecodeDataBlocks(data [][]byte) error {
+	if err := s.erasure.Reconstruct(data); err != nil {
+		return Errorf("failed to reconstruct data: %v", err)
+	}
+	return nil
+}
+
+// ErasureDecodeDataAndParityBlocks decodes the given erasure-coded data and verifies it.
+// It returns an error if the decoding failed.
+func (s *XLStorage) ErasureDecodeDataAndParityBlocks(data [][]byte) error {
+	if err := s.erasure.Reconstruct(data); err != nil {
+		return Errorf("failed to reconstruct data: %v", err)
+	}
+	if ok, err := s.erasure.Verify(data); !ok || err != nil {
+		if err != nil {
+			return Errorf("failed to verify data: %v", err)
```

Comment:
Because there are two different error conditions:
1. for example if we pass two few shards -> ErrTooFewShards
2. the data verification failed because e.g. the data is (after decode) is not correct.
If we remove one check we return the "wrong" error...

We need no !ok because if we enter the first if (`!ok || err != nil`)  but not the second (`err != nil`)
this implies that `ok == false`. 
Or didn't I get your point...